### PR TITLE
[CIS-2215] Implement UI tests attachments

### PR DESCRIPTION
**Pull request by alteral**
### 🔗 Issue Links

- https://stream-io.atlassian.net/browse/CIS-2215

### 🎯 Goal

- Create a way to attach logs and video recordings of the failed tests to have a better idea of the failures

### 📝 Summary

- Video attachments are collected via custom implementation using `sinatra`, the same way as we automated push notification testing
- Log attachments are collected via `xcparse` tool
- Both logs and videos can be found in the `Summary` section of the failed Github Actions run
- Small refactoring was done of the testing framework and the mock server (required for SwiftUI repo)
- Bump retry count due to `CIS-1737`, unfortunately, it's getting flakier
- Github Actions warnings were resolved:
    - `cache` and `checkout` actions versions were bumped
    - `set-output` method was transformed to `$GITHUB_OUTPUT` variable

### 🎨 Showcase

Screen Shot 2022-10-21 at 11 43 38 AM
Screen Shot 2022-10-21 at 11 52 07 AM

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [ ] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

gif

### CODE REVIEW

File: TestTools/StreamChatTestMockServer/MockServer/ChannelResponses.swift [hunk 1]
```diff
@@ -67,6 +67,10 @@ public extension StreamMockServer {
         server.register(MockEndpoint.truncate) { [weak self] request in
             self?.channelTruncation(request)
         }
+        server.register(MockEndpoint.sync) { request in
+            // TODO:
```
Comment (martinmitrevski) [inline TestTools/StreamChatTestMockServer/MockServer/ChannelResponses.swift:71]: Maybe we can write here what needs to be done?

File: StreamChatUITestsAppUITests/Tests/Base TestCase/StreamTestCase.swift [hunk 2]
```diff
@@ -14,24 +14,25 @@ class StreamTestCase: XCTestCase {
     var backendRobot: BackendRobot!
     var participantRobot: ParticipantRobot!
     var server: StreamMockServer!
+    let recordVideo = false
```
Comment (martinmitrevski) [inline StreamChatUITestsAppUITests/Tests/Base TestCase/StreamTestCase.swift:17]: How can we set the `recoredVideo` to true? (since it's a constant)

**Review by martinmitrevski — APPROVED** [review]
Great stuff @alteral 👏 Have two small comments, but nothing blocking.

File: TestTools/StreamChatTestMockServer/MockServer/ChannelResponses.swift (hunk 1 above)
Comment (alteral) [reply TestTools/StreamChatTestMockServer/MockServer/ChannelResponses.swift]: Done 👍

File: StreamChatUITestsAppUITests/Tests/Base TestCase/StreamTestCase.swift (hunk 2 above)
Comment (alteral) [reply StreamChatUITestsAppUITests/Tests/Base TestCase/StreamTestCase.swift]: Nice catch! Updated 👍

File: fastlane/sinatra.rb [hunk 3]
```diff
@@ -1,7 +1,37 @@
...
+      break unless File.exist?(video_file)
+    end
+  end
+
+  if body['stop']
+    simctl_processes = `pgrep simctl`.strip.split("\n")
+    simctl_processes.each { |pid| `kill -s SIGINT #{pid}` }
+    File.delete(video_file) if body['delete'] && File.exist?(video_file)
+  else
+    puts `xcrun simctl io #{params['udid']} recordVideo --codec h264 --force #{video_file} &`
```
Comment (polqf) [inline fastlane/sinatra.rb:35]: 👏 👏

**Review by polqf — APPROVED** [review]

**Review by nuno-vieira — APPROVED** [review]
LGTM! 🚀

This is really awesome! @alteral Can we also do this with failed snapshots? 🤔 It would be really nice to upload the failed snapshots to this folder as well. This is actually a long-lasting ticket we have: https://stream-io.atlassian.net/browse/CIS-651

**Comment by alteral** [comment]
> LGTM! 🚀
> 
> This is really awesome! @alteral Can we also do this with failed snapshots? 🤔 It would be really nice to upload the failed snapshots to this folder as well. This is actually a long-lasting ticket we have: https://stream-io.atlassian.net/browse/CIS-651

Thanks @nuno-vieira! Uploading failed snapshots is a bit more complicated than it sounds, but I'm pretty sure I'll come up with something soon enough 🙂

**Pull request merged**